airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -102,85 +102,247 @@ def _get_dep_statuses(
         if ti.task.trigger_rule == TR.ALWAYS:
             yield self._passing_status(reason="The task had a always trigger rule set.")
             return
+        if not ti.task.is_teardown:
+            # a teardown cannot have any indirect setups
+            relevant_setups = {t.task_id: t for t in ti.task.get_upstreams_only_setups()}
+            if relevant_setups:
+                for status, changed in self._evaluate_setup_constraint(
+                    ti=ti, relevant_setups=relevant_setups, dep_context=dep_context, session=session
+                ):
+                    yield status
+                    if not status.passed and changed:
+                        # no need to evaluate trigger rule; we've already marked as skipped or failed
+                        return
         yield from self._evaluate_trigger_rule(ti=ti, dep_context=dep_context, session=session)
 
-    def _evaluate_trigger_rule(
+    @staticmethod
+    @functools.lru_cache
+    def _get_relevant_upstream_map_indexes(*, ti: TaskInstance, upstream_task, session) -> int | range | None:
+        """Get the given task's map indexes relevant to the current ti.
+
+        This extra closure allows us to query the database only when needed,
+        and at most once for each task (instead of once for each expanded
+        task instance of the same task).
+        """
+        from airflow.models.abstractoperator import NotMapped
+        from airflow.models.expandinput import NotFullyPopulated
+
+        @functools.lru_cache
+        def _get_expanded_ti_count() -> int:
```

Comment:
And this cache has no effect at all since it is local to this function and only accessed exactly once during its lifetime.